### ISSUE TITLE
Mura not encoding HTML - being identified as SQL and cross site scripting attacks

### ISSUE THREAD

**Issue by JesseEarley** [blueriver/MuraCMS#2963]
Hello,

Our websites are now behind a university web application firewall that has rules in place to prevent sql and cross site scripting attacks. Currently a number of our sites are not able to publish because of these rules. In talking with university IT, it appears that Mura is **not** encoding the HTML that it creates and posts when publishing. The university policy in place won't be changing. Is HTML encoding something that can be added to Mura?

**Comment by mattlevine** [comment]
We take security very seriously.  Do you have any specific examples that you can share that are part of Mura's core codebase?  Perhaps send them to me privately at matt.levine@blueriver.com.  

Yes, Mura does store HTML is select database fields.  Primarily in tcontent.summary and tcontent.body.  Off the top of my head that may be it.   However, Mura properly escapes all sql queries. If you have found one where it doesn't please let me know which one it is and we will fix it ASAP.    If you use MuraORM you really can't avoid having all sql params sanitized.

In addition Mura has a built in script protection filter which sanitizes incoming Form, URL, CGI and Cookie data.

https://github.com/blueriver/MuraCMS/blob/7.1/core/mura/Portcullis.cfc

There are however by default some attributes that are not scanned because they are used by many people to add javascript that they intentionally want to run in the browser as part of their content. The default of these are here:

https://github.com/blueriver/MuraCMS/blob/7.1/core/mura/configBean.cfc#L128

You can easily customized this settings by adding this to your config/settings.ini.cfm and reloading:

scriptProtectExceptions=

Or adding an environment variable named MURA_SCRIPTPROTECTEXCEPTIONS and restart your service.

Any attribute specified should be treated as extra specially hostile. However, when you think of it, still like any other variable.  It should be parametrized when used within a sql query and sanitized when rendered, except for in a use case that you have explicitly design it for.  Like making the business decision to allow users to add javascript to their content bodies.  However, as stated you can turn that off.

This next part isn't necessarily for you Jesse, but I like to get it out there for people in general....

One important thing to understand though is what Mura does and doesn't do.  Mura is a platform on which people build their sites.  Mura is not responsible for how people build their sites on Mura.  I'm not saying that the issues you are seeing are certainly not Mura's fault.  I'm just explaining that things can go wrong security-wise in many ways that are simply out of our control. 

Here's a simple Mura lock-down guide that you can follow that should at least get you to a decent standard.

https://github.com/blueriver/mura-lockdown/blob/master/README.md

If you look at the lock-down guide you'll see that most of the items are simply making sure to sanitize all user and environment variables when rendering html and sql parameters.  If your organization has had developers not follow these basic rules then yes, you may have some issues.

So basically what you should do is identify where the code that is causing the issue is being rendered.  If it's in Mura's core please send it to me privately and we will get it patched ASAP.  However, if it's in code written by your organization, just reference the lockdown guide in how to deal with it.

-Matt
